test/util/framework/identities_helper.go

```diff
@@ -0,0 +1,660 @@
+		parameters: map[string]interface{}{},
+	}
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	if cfg.scope == BicepDeploymentScopeResourceGroup {
+		return nil, fmt.Errorf("DeployManagedIdentities cannot be called with BicepDeploymentScopeResourceGroup")
+	}
+
+	if cfg.deploymentName == "" {
+		cfg.deploymentName = fmt.Sprintf("mi-%s", cfg.resourceGroup)
+	}
+
+	usePooled := tc.UsePooledIdentities()
+
+	var msiRGName string
+	var identities Identities
+
+	if usePooled {
+		msiPool, err := tc.GetLeasedIdentities()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get leased MSIs: %w", err)
+		}
+		msiRGName = msiPool.ResourceGroupName
+		identities = msiPool.Identities
+	} else {
+		msiRGName = cfg.resourceGroup
+		identities = NewDefaultIdentities()
+	}
+
+	parameters := map[string]interface{}{
+		"clusterResourceGroupName": cfg.resourceGroup,
+		"msiResourceGroupName":     msiRGName,
+		"useMsiPool":               usePooled,
+		"identities":               identities,
+		"nsgName":                  cfg.parameters["nsgName"],
+		"vnetName":                 cfg.parameters["vnetName"],
+		"subnetName":               cfg.parameters["subnetName"],
+		"keyVaultName":             cfg.parameters["keyVaultName"],
+	}
+
+	deploymentResult, err := tc.CreateBicepTemplateAndWait(ctx,
+		WithTemplateFromBytes(cfg.template),
+		WithScope(cfg.scope),
+		WithDeploymentName(cfg.deploymentName),
+		WithLocation(invocationContext().Location()),
+		WithParameters(parameters),
+		WithTimeout(cfg.timeout),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create managed identities: %w", err)
+	}
+
+	return deploymentResult, nil
+}
+
+func (tc *perItOrDescribeTestContext) CreateIdentitiesPoolStateFile() error {
+
+	f, err := os.OpenFile(tc.msiPoolStateFilePath(), os.O_RDWR|os.O_CREATE, 0666)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	state := LeasedIdentityPoolState{file: f}
+
+	if err := state.lock(); err != nil {
+		return fmt.Errorf("failed to acquire managed identities pool state file lock: %w", err)
+	}
+	defer func() {
+		if err := state.unlock(); err != nil {
+			log.Logger.WithError(err).Warn("failed to release managed identities pool state file lock")
+		}
+	}()
+
+	if err := state.readUnlocked(); err != nil {
+		return fmt.Errorf("failed to read managed identities pool state file: %w", err)
+	}
+
+	if state.isInitialized() {
+		return nil
+	}
+
+	leasedRGs := tc.perBinaryInvocationTestContext.LeasedIdentityContainers()
+	if len(leasedRGs) == 0 {
+		return fmt.Errorf("expected envvar %s to not be empty", LeasedMSIContainersEnvvar)
+	}
+
+	if err := state.initializeUnlocked(leasedRGs); err != nil {
+		return fmt.Errorf("failed to initialize managed identities pool state: %w", err)
+	}
+
+	return nil
+}
+
+func (tc *perItOrDescribeTestContext) AssignIdentityContainers(ctx context.Context, count uint8, waitBetweenRetries time.Duration) error {
+	state, err := NewLeasedIdentityPoolState(tc.msiPoolStateFilePath())
```

Comment:
Why do we allow creating many instances of this when we use flock under the hook to limit concurrency anyway? A singleton in the `perInvocation` explains the intent better

Reply:
Moved to a singleton in the perInvocationTestContext. I also removed the state file initialization with `specs.AddBeforeAll()` as that was running once per process anyway. The instance is created once per process in the perInvocationTestContext and state file initialization is multi-process safe.